src/infrastructure/amqp/AbstractConsumer.ts

```diff
@@ -0,0 +1,168 @@
+import type { Either } from '@lokalise/node-core'
+import type { Channel, Connection, Message } from 'amqplib'
+import type { ZodSchema } from 'zod'
+
+import type { Dependencies } from '../diConfig'
+import { globalLogger, resolveGlobalErrorLogObject } from '../errors/globalErrorHandler'
+
+import type { ConsumerErrorResolver } from './ConsumerErrorResolver'
+import type { CommonMessage } from './MessageTypes'
+import { AmqpMessageInvalidFormat, AmqpValidationError } from './amqpErrors'
+import { deserializeMessage } from './messageDeserializer'
+
+export interface Consumer {
+  consume(): void
+  close(): Promise<void>
+}
+
+const ABORT_EARLY_EITHER: Either<'abort', never> = {
+  error: 'abort',
+}
+
+export type ConsumerParams<MessagePayloadType extends CommonMessage> = {
+  queueName: string
+  messageSchema: ZodSchema<MessagePayloadType>
+}
+
+export abstract class AbstractConsumer<MessagePayloadType extends CommonMessage>
+  implements Consumer
+{
+  protected readonly queueName: string
+  protected readonly connection: Connection
+  // @ts-ignore
+  protected channel: Channel
+  protected readonly errorHandler: ConsumerErrorResolver
+  private readonly newRelicBackgroundTransactionManager
+  private isShuttingDown: boolean
+
+  private messageSchema: ZodSchema<MessagePayloadType>
+
+  constructor(
+    params: ConsumerParams<MessagePayloadType>,
+    { amqpConnection, consumerErrorResolver, newRelicBackgroundTransactionManager }: Dependencies,
+  ) {
+    this.connection = amqpConnection
+    this.errorHandler = consumerErrorResolver
+    this.isShuttingDown = false
+    this.queueName = params.queueName
+    this.messageSchema = params.messageSchema
+    this.newRelicBackgroundTransactionManager = newRelicBackgroundTransactionManager
+  }
+
+  private async init() {
```

Comment:
:large_orange_diamond: When and where this method should be invoked? I see there is `consume` method but when it is supposed to be called? Is `server` or `app` a good place for that?

Reply:
It's a private method, and yes, it's invoked early - `app` does it now, and in consumer tests we also do explicit destroy/init for the cleanup.